onnxruntime/core/providers/cuda/nn/pool.cc

```diff
@@ -182,16 +182,20 @@ Status Pool<T, PoolType, NHWC>::ComputeInternal(OpKernelContext* context) const
     if (NHWC) {
       x_dims_cudnn.insert(x_dims_cudnn.begin() + 1, 1);
       y_dims_cudnn.insert(y_dims_cudnn.begin() + 1, 1);
+      ORT_ENFORCE(pads.size() >= kernel_shape.size());
```

Comment:
TODO: I'm not sure what's the correct order for this NHWC path.

Reply:
The NHWC implementation seems not right: `x_dims_cudnn.insert(x_dims_cudnn.begin() + 1, 1)` means it adds a dimension H (like NWC => NHWC). Then the kernel_shape and pads shall pad to the first instead of second.

I think a proper implementation (NHC => NHWC):

if (NHWC) {
      x_dims_cudnn.insert(x_dims_cudnn.begin() + 2, 1);
      y_dims_cudnn.insert(y_dims_cudnn.begin() + 2, 1);
      ORT_ENFORCE(pads.size() >= kernel_shape.size());
      pads.insert(pads.begin() + kernel_shape.size(), 0);
      pads.insert(pads.end(), 0);
      kernel_shape.push_back(1);
      strides.push_back(1);
 }
 
 Then we can refactor it share last 5 lines of common code for NHWC and NCHW.

Reply:
Agree. We have been adding the extra feature dim to the right of the existing feature dim - not the to the left. 

(1) It seems like - we need 1-D NHWC Pool tests added (would have caught the issue Tianlei mentioned)
(2) The existing 1-D NCHW Pool tests need some enhancements -  (https://github.com/microsoft/onnxruntime/blob/06e684c9f2f8495de5259967cc12bab24da3d522/onnxruntime/test/providers/cpu/nn/pool_op_test.cc#L202) seem to only have tests with pads as 0s meaning it can't catch this bug. If we had a 1-D test with pads like {0,1} for example - it would have made pads {0,1,0,0} with this change instead of the correct {0,0,1,0} with the previous logic.

Reply:
Will update the code shortly.

Reply:
> BTW, the code for global pooling also need add NHWC support (in a separated PR):

Right. But how could we already have a unit test for it, and it is passing? See this commit: https://github.com/microsoft/onnxruntime/pull/19780/commits/8e1b4b8c8692893f694f06880452440dcd2013c1 
The commit was trying to disable the test.